rust/kernel/of.rs

```diff
@@ -4,72 +4,84 @@
+where
+    [bindings::of_device_id; N + 1]: Sized,
+{
+    /// Create a new Open Firmware Match Table from a list of compatible strings.
+    pub const fn new_const(compatibles: [&'static CStr; N]) -> Self {
+        let mut ids = [Self::zeroed_of_device_id(); N + 1];
+        let mut i = 0;
+        loop {
+            if i == N {
+                break;
+            }
+            ids[i] = Self::new_of_device_id(compatibles[i]);
+            i += 1;
+        }
+        Self(ids)
     }
 
-    fn new_of_device_id(compatible: &'static CStr) -> Result<bindings::of_device_id> {
-        let mut buf = [0_u8; 128];
-        if compatible.len() > buf.len() {
-            return Err(Error::EINVAL);
+    const fn zeroed_of_device_id() -> bindings::of_device_id {
+        bindings::of_device_id {
+            name: [0; 32usize],
+            type_: [0; 32usize],
+            compatible: [0; 128usize],
+            data: ptr::null(),
         }
-        buf.get_mut(..compatible.len())
-            .ok_or(Error::EINVAL)?
-            .copy_from_slice(compatible.as_bytes());
-        Ok(bindings::of_device_id {
-            // SAFETY: re-interpretation from [u8] to [c_types::c_char] of same length is always safe.
-            compatible: unsafe { transmute::<[u8; 128], [c_types::c_char; 128]>(buf) },
-            ..Default::default()
-        })
     }
-}
 
-impl PointerWrapper for OfMatchTable {
-    fn into_pointer(self) -> *const c_types::c_void {
-        // Per the invariant above, the generated pointer points to an
-        // array of `bindings::of_device_id`, where the final element is
-        // filled with zeros (the sentinel). Therefore, it's suitable to
-        // be assigned to `bindings::device_driver::of_match_table`.
-        self.0.into_pointer()
+    const fn new_of_device_id(compatible: &'static CStr) -> bindings::of_device_id {
+        let mut id = Self::zeroed_of_device_id();
+        let compatible = compatible.as_bytes_with_nul();
+        let mut i = 0;
+        loop {
+            if i == compatible.len() {
+                break;
+            }
+            // if `compatible` does not fit in `id.compatible`, an
+            // "index out of bounds" build time exception will be triggered.
+            id.compatible[i] = compatible[i] as c_types::c_char;
+            i += 1;
+        }
+        id
     }
+}
 
-    unsafe fn from_pointer(p: *const c_types::c_void) -> Self {
-        // SAFETY: The passed pointer comes from a previous call to [`InnerTable::into_pointer()`].
-        Self(unsafe { InnerTable::from_pointer(p) })
+impl<const N: usize> OfMatchTable for ConstOfMatchTable<N>
+where
+    [bindings::of_device_id; N + 1]: Sized,
+{
+    fn as_ptr(&'static self) -> *const bindings::of_device_id {
+        // INVARIANT: the returned pointer is created by dereferencing
+        // a structure with static lifetime, therefore the pointer itself
+        // has static lifetime.
+        &self.0[0]
```

Comment:
@alex @nbdd0121  is a `*const` pointer to
```rust
[bindings::of_device_id; N]
```
always equivalent to:
```c
struct of_device_id ids[N];
```
? Or do I need a `repr` annotation somewhere? What I'm doing here does appear to work on at least two platforms, but stranger things have happened.

Reply:
They are. Given arrays are primitives you don't have a place to put a `repr`.

I suspect the strange thing you observe is related to the usage of incomplete features.